### DISCUSSION TITLE
How to move focus to next control

### DISCUSSION THREAD

**Discussion by mysteryx93** [AvaloniaUI/Avalonia#7603] [Q&A] [answered]
What's the equivalent of this line in Avalonia?

    control.MoveFocus(new TraversalRequest(FocusNavigationDirection.Next));

**Comment by amwx** [comment, marked as answer]
```C#
var current = FocusManager.Instance?.Current;
if (current != null)
{
    var next = KeyboardNavigationHandler.GetNext(current, NavigationDirection.Next);
    if (next != null)
    {
        FocusManager.Instance?.Focus(next, NavigationMethod.Directional);
    }
}
```

Note that `GetNext()` typically only accepts `NavigationDirection.Next` or `NavigationDirection.Previous`. This also just pulls the next focusable element it finds, and doesn't do any comprehensive spatial searching

**Reply by pdesterlich** [reply]
With latest version of avalonia this code has 2 problems:
- `FocusManager.Instance` fails, there's no `Instance` property in `FocusManager`
- `KeyboardNavigationHandler.GetNext` is marked as Obsolete

the first can be replaced by `FocusManager?.GetFocusedElement()`, but what can be used as an alternative to `GetNext` ?

**Reply by longbuibao** [reply]
hi, anyone have update on this?

**Reply by maxkatz6** [reply]
`FocusManager` is a property on TopLevel/Window. See https://docs.avaloniaui.net/docs/concepts/services/focus-manager.

`KeyboardNavigationHandler.GetNext` is obsolete, but still available for now.
For 12.0 release we going to have new API, with this PR: https://github.com/AvaloniaUI/Avalonia/pull/18647

**Comment by mysteryx93** [comment]
Great that works!

Except for ListBox which doesn't respect TabIndex, but that's a separate issue so I'll create another thread.
